bundles/org.openhab.binding.panamaxfurman/src/main/java/org/openhab/binding/panamaxfurman/internal/transport/PanamaxFurmanSocket.java

```diff
@@ -0,0 +1,203 @@
+
+        @Override
+        public void run() {
+            try {
+                logger.debug("InputStreamReaderThread.run stopReader={}  @{}", stopReader, connectionName);
+                while (!stopReader && !Thread.currentThread().isInterrupted()) {
+                    String receivedData = null;
+                    try {
+                        logger.trace("InputStreamReaderThread.run waiting to read data  @{}", connectionName);
+                        // Blocks forever until data is received so DO NOT synchronize
+                        receivedData = bufferedReader.readLine();
+                        if (receivedData == null) {
+                            throw new IOException("End of stream reached " + connectionName);
+                        } else {
+                            logger.trace("received: {}    @{}", receivedData, connectionName);
+                            if (receivedData.trim().length() > 0) {
+                                if (!logger.isTraceEnabled()) {
+                                    logger.debug("received: {}    @{}", receivedData, connectionName);
+                                }
+                                eventListener.onInformationReceived(receivedData);
+                            }
+                        }
+                    } catch (SocketTimeoutException | SocketException e) {
+                        if (e instanceof SocketException && e.getMessage() != null
+                                && !e.getMessage().contains("Connection timed out")) {
+                            throw e;
+                        }
+                        logger.info("{} while trying to read data, retrying.  detail: {}.  @{}",
+                                e.getClass().getSimpleName(), e.getMessage(), connectionName);
+                        // Let the while loop try again
+                    }
+                }
+            } catch (IOException e) {
+                // This is normal when the socket is closed, so don't log anything if we are stopping
+                if (!stopReader) {
+                    logger.warn("Error reading data from @{}. Cause: {}  ", connectionName, e.getMessage(),
+                            Util.exceptionToLog(logger, e));
+                    eventListener.onConnectivityEvent(new PanamaxFurmanConnectionBrokenEvent(e.getMessage()));
+                }
+            } finally {
+                // Notify the stopReader method caller that the reader is stopped.
+                this.stopLatch.countDown();
+                Util.closeQuietly(bufferedReader);
+            }
+            logger.debug("InputStreamReaderThread thread exiting.  @{}", connectionName);
+        }
+
+        private void shutdown() {
+            stopReader = true;
+            Util.closeQuietly(bufferedReader);
+            try {
+                // This needs to return quickly so keep the timeout short
+                boolean stopped = this.stopLatch.await(500, TimeUnit.MILLISECONDS);
```

Comment:
Think something for a maintainer to look at, as i think 500ms is not quick, but don't have a better solution at the moment.

Reply:
If it is not closed quietly, what will happen after the timeout is hit. Would it cause a leak, or what will happen with the resource?

Reply:
Good question, let me do some research and some testing on this one.  Will circle back on the 500ms timeout too